cli/cmd/metrics_diagnostics_util.go

```diff
@@ -0,0 +1,150 @@
+package cmd
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"sort"
+	"sync/atomic"
+	"time"
+
+	"github.com/linkerd/linkerd2/pkg/k8s"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// shared between metrics and diagnostics command
+type metricsResult struct {
+	pod       string
+	container string
+	metrics   []byte
+	err       error
+}
+type byResult []metricsResult
+
+func (s byResult) Len() int {
+	return len(s)
+}
+func (s byResult) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+func (s byResult) Less(i, j int) bool {
+	return s[i].pod < s[j].pod || ((s[i].pod == s[j].pod) && s[i].container < s[j].container)
+}
+
+// getResponse makes a http Get request to the passed url and returns the response/error
+func getResponse(url string) ([]byte, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	bytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return bytes, nil
+}
+
+// getContainerMetrics returns the metrics exposed by a container on the passed in portName
+func getContainerMetrics(
+	k8sAPI *k8s.KubernetesAPI,
+	pod corev1.Pod,
+	container corev1.Container,
+	emitLogs bool,
+	portName string,
+) ([]byte, error) {
+	portForward, err := k8s.NewContainerMetricsForward(k8sAPI, pod, container, emitLogs, portName)
+	if err != nil {
+		return nil, err
+	}
+
+	defer portForward.Stop()
+	if err = portForward.Init(); err != nil {
+		fmt.Fprintf(os.Stderr, "Error running port-forward: %s", err)
+	}
+
+	metricsURL := portForward.URLFor("/metrics")
+	bytes, err := getResponse(metricsURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return bytes, nil
+}
+
+// getAllContainersWithPort returns all the containers within
+// a pod which exposes metrics at a port with name portName
+func getAllContainersWithPort(
+	pod corev1.Pod,
+	portName string,
+) ([]corev1.Container, error) {
+	if pod.Status.Phase != corev1.PodRunning {
+		return nil, fmt.Errorf("pod not running: %s", pod.GetName())
+	}
+	var containers []corev1.Container
+
+	for _, c := range pod.Spec.Containers {
+		for _, p := range c.Ports {
+			if p.Name == portName {
+				containers = append(containers, c)
+			}
+		}
+	}
+	return containers, nil
+}
+
+// getMetrics returns the metrics exposed by all the containers of the passed in list of pods
+// which exposes their metrics at portName
+func getMetrics(
+	k8sAPI *k8s.KubernetesAPI,
+	pods []corev1.Pod,
+	portName string,
+	waitingTime time.Duration,
+	emitLogs bool,
+) []metricsResult {
+	var results []metricsResult
+
+	resultChan := make(chan metricsResult)
+	var activeRoutines int32
+	for _, pod := range pods {
+		containers, err := getAllContainersWithPort(pod, portName)
+		if err != nil {
+			fmt.Println(err)
```

Comment:
Generally, it's better to have a consistent error-handling mechanism. Is it possible to wrap this in a `metricsResult` instance, containing just the `pod` and `err` fields, and append it to the result slice?

Reply:
In this case, we'll have to do something like this in `cli/cmd/diagnostics`

```golang
content := fmt.Sprintf("#\n# POD %s (%d of %d)\n", result.pod, i+1, len(results))
if result.container != "" {
    content += fmt.Sprintf("# CONTAINER %s \n#\n", result.container)
}
```

Reply:
Sure :+1:. Or if you want to reduce the if/else, you can do:

```diff
diff --git a/cli/cmd/diagnostics.go b/cli/cmd/diagnostics.go
index 70af43fc..eed45d1c 100644
--- a/cli/cmd/diagnostics.go
+++ b/cli/cmd/diagnostics.go
@@ -47,13 +47,11 @@ func newCmdDiagnostics() *cobra.Command {
 			var buf bytes.Buffer
 			for i, result := range results {
 				content := fmt.Sprintf("#\n# POD %s (%d of %d)\n", result.pod, i+1, len(results))
-				if result.container != "" {
+				if result.err != nil {
+					content += fmt.Sprintf("# ERROR %s\n", result.err)
+				} else {
 					content += fmt.Sprintf("# CONTAINER %s \n#\n", result.container)
-				}
-				if result.err == nil {
 					content += string(result.metrics)
-				} else {
-					content += fmt.Sprintf("# ERROR %s\n", result.err)
 				}
 				buf.WriteString(content)
 			}
```

Reply:
done